### PR TITLE
UTF-16 + Unicode file path support in Windows

### PR DESCRIPTION
**Pull request by elasota**
Adds UTF-8/UTF-16 conversion and uses that to support Unicode file paths in Windows.

### CODE REVIEW

File: src/c/include/pn/procyon.h [hunk 1]
```diff
@@ -24,14 +24,15 @@
...
-typedef struct pn_string  pn_string_t;
-typedef struct pn_array   pn_array_t;
-typedef struct pn_map     pn_map_t;
-typedef struct pn_kv_pair pn_kv_pair_t;
-typedef struct pn_input   pn_input_t;
-typedef struct pn_output  pn_output_t;
+typedef struct pn_value     pn_value_t;
+typedef struct pn_data      pn_data_t;
+typedef struct pn_string    pn_string_t;
+typedef struct pn_string16  pn_string16_t;
```
Comment (sfiera) [inline src/c/include/pn/procyon.h]: I would rather not have two different string types in the public interface of this library. It seems to be orthogonal to the actual purpose of this change (supporting UTF-16 in `pn_path_input()` / `pn_path_output()`).

On the other hand, it would probably be useful to have string constructors from `wchar_t*` / `std::wstring`.

File: src/c/src/utf16.h [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+#define PROCYON_UTF16_H_
+
+#include <pn/procyon.h>
+#include <sys/types.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif  // __cplusplus
+
+void pn_u16chr(pn_rune_t rune, uint16_t* data, size_t* count);
```
Comment (sfiera) [inline src/c/src/utf16.h]: Seems a little silly to have two single-function headers `utf8.h` and `utf16.h`. Maybe `unicode.h` / `unicode.c` with both UTF-8 and UTF-16?

File: src/c/src/utf16.c [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+
+        in_offset = pn_rune_next(src, len, in_offset);
+        pn_u16chr(rune, str16->values + out_offset, &out_size);
+        out_offset += out_size;
+    }
+
+    return str16;
+}
+
+pn_string_t* pn_string_from_utf8(const uint16_t* src, size_t len) {
```
Comment (sfiera) [inline src/c/src/utf16.c]:
```suggestion
pn_string_t* pn_string_from_utf16(const uint16_t* src, size_t len) {
```

File: src/c/src/utf16.c [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+    if (is_low_surrogate(data[index])) {
+        return index + 1;
+    }
+    if (!is_high_surrogate(data[0])) {
+        return index + 1;
+    }
+    if (size < 2 || !is_low_surrogate(data[1])) {
+        return index + 1;
+    }
+    return index + 2;
```
Comment (sfiera) [inline src/c/src/utf16.c]: This could be much simpler than the UTF-8 version:
```suggestion
    if (size >= 2 && is_high_surrogate(data[0]) && is_low_surrogate(data[1])) {
        return index + 2;
    }
    return index + 1;
```

File: src/c/src/file.c [hunk 5]
```diff
@@ -22,12 +22,34 @@
...
+#ifdef _WIN32
+#define PN_FOPEN_MODE(s) L ## s
+#else
+#define PN_FOPEN_MODE(s) s
+#endif
+
+
+#ifdef _WIN32
+static FILE* pn_fopen_utf8(const char* path, const wchar_t* mode) {
+    pn_string16_t* wpath = pn_string16_from_utf8(path, strlen(path));
```
Comment (sfiera) [inline src/c/src/file.c]: There doesn’t seem to be an include for `common.h` from this file. WINE is failing but this seems to be just a warning in the native Windows build:
```
../../../src/c/src/file.c(34): warning C4047: 'initializing': 'pn_string16_t *' differs in levels of indirection from 'int'
```

File: src/c/include/pn/procyon.h (hunk 1 above)
Comment (elasota) [reply src/c/include/pn/procyon.h]: Yeah not sure where this should live... problem is it's needed in libsfz too.  Maybe it'd be better to move this to a separate library.

Problem with wchar_t is it's implementation-defined, it's UTF-32 on Linux despite not being useful there, there are still no standard encode functions in C, and the C++ ones are being revised.  So, it really only makes sense to implement UTF-16 for dealing with file paths.

Comment (sfiera) [reply src/c/include/pn/procyon.h]: Could you use `char16_t` / `std::u16string` for Windows? libsfz can assume C++11 support.

**Comment by elasota** [comment]
I'll rework this when #38 is merged I guess.

**Comment by sfiera** [comment]
It’s merged. Sorry for the wait. I ended up disabling the WINE build; I’m not sure exactly where it was going wrong and with a native Windows build it’s not so important.

**Comment by elasota** [comment]
Reworked this to just deal with the fopen in light of other changes, so this should be ready for review/merge now.

**Comment by sfiera** [comment]
Looks good, thanks.
